module/zfs/dmu_recv.c

```diff
@@ -2558,6 +2589,173 @@ receive_process_record(struct receive_writer_arg *rwa,
 	return (err);
 }
 
+static int
+receive_heal_record(struct receive_writer_arg *rwa,
+    struct receive_record_arg *rrd)
+{
+	uint64_t size, lsize = 0, csize = 0, blkid = 0, obj = ZB_ROOT_OBJECT;
+	dmu_buf_t *dbp;
+	abd_t *abd;
+	void *buf;
+	int err = 0;
+	blkptr_t *bp = NULL;
+	objset_t *os = rwa->os;
+	dmu_replay_record_t drr = rrd->header;
+	enum zio_flag flags = ZIO_FLAG_SPECULATIVE |
+	    ZIO_FLAG_DONT_CACHE | ZIO_FLAG_DONT_RETRY | ZIO_FLAG_CANFAIL;
+
+	/* Should only get here if we are doing a corrective recv */
+	ASSERT(rwa->heal);
+	/* Processing in order, therefore bytes_read should be increasing. */
+	ASSERT3U(rrd->bytes_read, >=, rwa->bytes_read);
+
+	rwa->bytes_read = rrd->bytes_read;
+
+	/* We can only heal write and spill records; other ones get ignored */
+	if (drr.drr_type != DRR_WRITE && drr.drr_type != DRR_SPILL)
+		goto cleanup;
+
+	switch (drr.drr_type) {
+	case DRR_WRITE:
+	{
+		uint64_t offset = drr.drr_u.drr_write.drr_offset;
+		obj = drr.drr_u.drr_write.drr_object;
+		lsize = drr.drr_u.drr_write.drr_logical_size;
+		csize = drr.drr_u.drr_write.drr_compressed_size;
+
+		buf = kmem_alloc(lsize, KM_SLEEP);
+		/* Try to read the object to see if it needs healing */
+		err = dmu_read(os, obj, offset, lsize, buf,
+		    DMU_READ_NO_PREFETCH);
+		kmem_free(buf, lsize);
+		if (err != ECKSUM)
+			goto cleanup; /* no corruption found */
+
+		err = dmu_buf_hold_noread(os, obj, offset, FTAG, &dbp);
+		if (err != 0) {
+			err = SET_ERROR(EBUSY);
+			goto cleanup;
+		}
+		blkid =
+		    dbuf_whichblock(DB_DNODE((dmu_buf_impl_t *)dbp), 0, offset);
```

Comment:
You need to use `DB_DNODE_ENTER/EXIT` around the `DB_DNODE()`, or better yet `dmu_buf_dnode_enter()` as you do below.  Or if you're going to cast the dbuf, you might as well just dereference `db_blkid`